File: krisideepweb/src/models/GovernmentScheme.ts
export interface GovernmentScheme {
  id: string;
  schemeName: string;
  schemeNameHindi: string;
  description: string;
  descriptionHindi: string;
  category: string; // subsidy, loan, insurance, welfare
  eligibilityCriteria: string;
  eligibilityCriteriaHindi: string;
  maxBenefit: number;
  benefitType: string; // amount, percentage, fixed
  requiredDocuments: string[];
  requiredDocumentsHindi: string[];
  applicationProcess: string;
  applicationProcessHindi: string;
  contactInfo: string;
  startDate: Date;
  endDate?: Date;
  isActive: boolean;
  state: string;
  stateHindi: string;
}

export interface SubsidyApplication {
  id: string;
  schemeId: string;
  farmerName: string;
  farmerId: string;
  mobileNumber: string;
  applicationData: Record<string, any>;
  status: string; // submitted, under_review, approved, rejected, disbursed
  submissionDate: Date;
  approvalDate?: Date;
  disbursalDate?: Date;
  approvedAmount?: number;
  rejectionReason?: string;
  rejectionReasonHindi?: string;
  uploadedDocuments: string[];
}

export interface EligibilityResult {
  schemeId: string;
  isEligible: boolean;
  eligibilityScore: number; // 0-100
  eligibleCriteria: string[];
  ineligibleCriteria: string[];
  eligibleCriteriaHindi: string[];
  ineligibleCriteriaHindi: string[];
  estimatedBenefit: number;
  recommendations: string[];
  recommendationsHindi: string[];
}

export interface SchemeNotification {
  id: string;
  title: string;
  titleHindi: string;
  message: string;
  messageHindi: string;
  type: string; // new_scheme, deadline_reminder, status_update, general
  schemeId?: string;
  createdAt: Date;
  expiryDate?: Date;
  isRead: boolean;
  isImportant: boolean;
}

export interface GovernmentSchemeJson {
  id: string;
  schemeName: string;
  schemeNameHindi: string;
  description: string;
  descriptionHindi: string;
  category: string;
  eligibilityCriteria: string;
  eligibilityCriteriaHindi: string;
  maxBenefit: number;
  benefitType: string;
  requiredDocuments: string[];
  requiredDocumentsHindi: string[];
  applicationProcess: string;
  applicationProcessHindi: string;
  contactInfo: string;
  startDate: string;
  endDate?: string;
  isActive: boolean;
  state: string;
  stateHindi: string;
}

export interface SubsidyApplicationJson {
  id: string;
  schemeId: string;
  farmerName: string;
  farmerId: string;
  mobileNumber: string;
  applicationData: Record<string, any>;
  status: string;
  submissionDate: string;
  approvalDate?: string;
  disbursalDate?: string;
  approvedAmount?: number;
  rejectionReason?: string;
  rejectionReasonHindi?: string;
  uploadedDocuments: string[];
}

export interface EligibilityResultJson {
  schemeId: string;
  isEligible: boolean;
  eligibilityScore: number;
  eligibleCriteria: string[];
  ineligibleCriteria: string[];
  eligibleCriteriaHindi: string[];
  ineligibleCriteriaHindi: string[];
  estimatedBenefit: number;
  recommendations: string[];
  recommendationsHindi: string[];
}

export interface SchemeNotificationJson {
  id: string;
  title: string;
  titleHindi: string;
  message: string;
  messageHindi: string;
  type: string;
  schemeId?: string;
  createdAt: string;
  expiryDate?: string;
  isRead: boolean;
  isImportant: boolean;
}

export const governmentSchemeFromJson = (json: GovernmentSchemeJson): GovernmentScheme => ({
  id: json.id,
  schemeName: json.schemeName,
  schemeNameHindi: json.schemeNameHindi,
  description: json.description,
  descriptionHindi: json.descriptionHindi,
  category: json.category,
  eligibilityCriteria: json.eligibilityCriteria,
  eligibilityCriteriaHindi: json.eligibilityCriteriaHindi,
  maxBenefit: json.maxBenefit,
  benefitType: json.benefitType,
  requiredDocuments: json.requiredDocuments,
  requiredDocumentsHindi: json.requiredDocumentsHindi,
  applicationProcess: json.applicationProcess,
  applicationProcessHindi: json.applicationProcessHindi,
  contactInfo: json.contactInfo,
  startDate: new Date(json.startDate),
  endDate: json.endDate ? new Date(json.endDate) : undefined,
  isActive: json.isActive ?? true,
  state: json.state,
  stateHindi: json.stateHindi,
});

export const subsidyApplicationFromJson = (json: SubsidyApplicationJson): SubsidyApplication => ({
  id: json.id,
  schemeId: json.schemeId,
  farmerName: json.farmerName,
  farmerId: json.farmerId,
  mobileNumber: json.mobileNumber,
  applicationData: json.applicationData,
  status: json.status,
  submissionDate: new Date(json.submissionDate),
  approvalDate: json.approvalDate ? new Date(json.approvalDate) : undefined,
  disbursalDate: json.disbursalDate ? new Date(json.disbursalDate) : undefined,
  approvedAmount: json.approvedAmount,
  rejectionReason: json.rejectionReason,
  rejectionReasonHindi: json.rejectionReasonHindi,
  uploadedDocuments: json.uploadedDocuments || [],
});

export const eligibilityResultFromJson = (json: EligibilityResultJson): EligibilityResult => ({
  schemeId: json.schemeId,
  isEligible: json.isEligible,
  eligibilityScore: json.eligibilityScore,
  eligibleCriteria: json.eligibleCriteria,
  ineligibleCriteria: json.ineligibleCriteria,
  eligibleCriteriaHindi: json.eligibleCriteriaHindi,
  ineligibleCriteriaHindi: json.ineligibleCriteriaHindi,
  estimatedBenefit: json.estimatedBenefit,
  recommendations: json.recommendations,
  recommendationsHindi: json.recommendationsHindi,
});

export const schemeNotificationFromJson = (json: SchemeNotificationJson): SchemeNotification => ({
  id: json.id,
  title: json.title,
  titleHindi: json.titleHindi,
  message: json.message,
  messageHindi: json.messageHindi,
  type: json.type,
  schemeId: json.schemeId,
  createdAt: new Date(json.createdAt),
  expiryDate: json.expiryDate ? new Date(json.expiryDate) : undefined,
  isRead: json.isRead ?? false,
  isImportant: json.isImportant ?? false,
});

export const governmentSchemeToJson = (scheme: GovernmentScheme): GovernmentSchemeJson => ({
  id: scheme.id,
  schemeName: scheme.schemeName,
  schemeNameHindi: scheme.schemeNameHindi,
  description: scheme.description,
  descriptionHindi: scheme.descriptionHindi,
  category: scheme.category,
  eligibilityCriteria: scheme.eligibilityCriteria,
  eligibilityCriteriaHindi: scheme.eligibilityCriteriaHindi,
  maxBenefit: scheme.maxBenefit,
  benefitType: scheme.benefitType,
  requiredDocuments: scheme.requiredDocuments,
  requiredDocumentsHindi: scheme.requiredDocumentsHindi,
  applicationProcess: scheme.applicationProcess,
  applicationProcessHindi: scheme.applicationProcessHindi,
  contactInfo: scheme.contactInfo,
  startDate: scheme.startDate.toISOString(),
  endDate: scheme.endDate?.toISOString(),
  isActive: scheme.isActive,
  state: scheme.state,
  stateHindi: scheme.stateHindi,
});

export const subsidyApplicationToJson = (application: SubsidyApplication): SubsidyApplicationJson => ({
  id: application.id,
  schemeId: application.schemeId,
  farmerName: application.farmerName,
  farmerId: application.farmerId,
  mobileNumber: application.mobileNumber,
  applicationData: application.applicationData,
  status: application.status,
  submissionDate: application.submissionDate.toISOString(),
  approvalDate: application.approvalDate?.toISOString(),
  disbursalDate: application.disbursalDate?.toISOString(),
  approvedAmount: application.approvedAmount,
  rejectionReason: application.rejectionReason,
  rejectionReasonHindi: application.rejectionReasonHindi,
  uploadedDocuments: application.uploadedDocuments,
});

export const eligibilityResultToJson = (result: EligibilityResult): EligibilityResultJson => ({
  schemeId: result.schemeId,
  isEligible: result.isEligible,
  eligibilityScore: result.eligibilityScore,
  eligibleCriteria: result.eligibleCriteria,
  ineligibleCriteria: result.ineligibleCriteria,
  eligibleCriteriaHindi: result.eligibleCriteriaHindi,
  ineligibleCriteriaHindi: result.ineligibleCriteriaHindi,
  estimatedBenefit: result.estimatedBenefit,
  recommendations: result.recommendations,
  recommendationsHindi: result.recommendationsHindi,
});

export const schemeNotificationToJson = (notification: SchemeNotification): SchemeNotificationJson => ({
  id: notification.id,
  title: notification.title,
  titleHindi: notification.titleHindi,
  message: notification.message,
  messageHindi: notification.messageHindi,
  type: notification.type,
  schemeId: notification.schemeId,
  createdAt: notification.createdAt.toISOString(),
  expiryDate: notification.expiryDate?.toISOString(),
  isRead: notification.isRead,
  isImportant: notification.isImportant,
});

File: krishideep_app/lib/models/government_scheme.dart
class GovernmentScheme {
  final String id;
  final String schemeName;
  final String schemeNameHindi;
  final String description;
  final String descriptionHindi;
  final String category; // subsidy, loan, insurance, welfare
  final String eligibilityCriteria;
  final String eligibilityCriteriaHindi;
  final double maxBenefit;
  final String benefitType; // amount, percentage, fixed
  final List<String> requiredDocuments;
  final List<String> requiredDocumentsHindi;
  final String applicationProcess;
  final String applicationProcessHindi;
  final String contactInfo;
  final String? applicationWebsite;
  final DateTime startDate;
  final DateTime? endDate;
  final bool isActive;
  final String state;
  final String stateHindi;

  GovernmentScheme({
    required this.id,
    required this.schemeName,
    required this.schemeNameHindi,
    required this.description,
    required this.descriptionHindi,
    required this.category,
    required this.eligibilityCriteria,
    required this.eligibilityCriteriaHindi,
    required this.maxBenefit,
    required this.benefitType,
    required this.requiredDocuments,
    required this.requiredDocumentsHindi,
    required this.applicationProcess,
    required this.applicationProcessHindi,
    required this.contactInfo,
    this.applicationWebsite,
    required this.startDate,
    this.endDate,
    this.isActive = true,
    required this.state,
    required this.stateHindi,
  });

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'schemeName': schemeName,
      'schemeNameHindi': schemeNameHindi,
      'description': description,
      'descriptionHindi': descriptionHindi,
      'category': category,
      'eligibilityCriteria': eligibilityCriteria,
      'eligibilityCriteriaHindi': eligibilityCriteriaHindi,
      'maxBenefit': maxBenefit,
      'benefitType': benefitType,
      'requiredDocuments': requiredDocuments,
      'requiredDocumentsHindi': requiredDocumentsHindi,
      'applicationProcess': applicationProcess,
      'applicationProcessHindi': applicationProcessHindi,
      'contactInfo': contactInfo,
      'applicationWebsite': applicationWebsite,
      'startDate': startDate.toIso8601String(),
      'endDate': endDate?.toIso8601String(),
      'isActive': isActive,
      'state': state,
      'stateHindi': stateHindi,
    };
  }

  factory GovernmentScheme.fromJson(Map<String, dynamic> json) {
    return GovernmentScheme(
      id: json['id'],
      schemeName: json['schemeName'],
      schemeNameHindi: json['schemeNameHindi'],
      description: json['description'],
      descriptionHindi: json['descriptionHindi'],
      category: json['category'],
      eligibilityCriteria: json['eligibilityCriteria'],
      eligibilityCriteriaHindi: json['eligibilityCriteriaHindi'],
      maxBenefit: json['maxBenefit'],
      benefitType: json['benefitType'],
      requiredDocuments: List<String>.from(json['requiredDocuments']),
      requiredDocumentsHindi: List<String>.from(json['requiredDocumentsHindi']),
      applicationProcess: json['applicationProcess'],
      applicationProcessHindi: json['applicationProcessHindi'],
      contactInfo: json['contactInfo'],
      applicationWebsite: json['applicationWebsite'],
      startDate: json['startDate'] != null
          ? DateTime.parse(json['startDate'])
          : DateTime.now(),
      endDate: json['endDate'] != null ? DateTime.parse(json['endDate']) : null,
      isActive: json['isActive'] ?? true,
      state: json['state'],
      stateHindi: json['stateHindi'],
    );
  }
}

class SubsidyApplication {
  final String id;
  final String schemeId;
  final String farmerName;
  final String farmerId;
  final String mobileNumber;
  final Map<String, dynamic> applicationData;
  final String status; // submitted, under_review, approved, rejected, disbursed
  final DateTime submissionDate;
  final DateTime? approvalDate;
  final DateTime? disbursalDate;
  final double? approvedAmount;
  final String? rejectionReason;
  final String? rejectionReasonHindi;
  final List<String> uploadedDocuments;

  SubsidyApplication({
    required this.id,
    required this.schemeId,
    required this.farmerName,
    required this.farmerId,
    required this.mobileNumber,
    required this.applicationData,
    required this.status,
    required this.submissionDate,
    this.approvalDate,
    this.disbursalDate,
    this.approvedAmount,
    this.rejectionReason,
    this.rejectionReasonHindi,
    this.uploadedDocuments = const [],
  });

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'schemeId': schemeId,
      'farmerName': farmerName,
      'farmerId': farmerId,
      'mobileNumber': mobileNumber,
      'applicationData': applicationData,
      'status': status,
      'submissionDate': submissionDate.toIso8601String(),
      'approvalDate': approvalDate?.toIso8601String(),
      'disbursalDate': disbursalDate?.toIso8601String(),
      'approvedAmount': approvedAmount,
      'rejectionReason': rejectionReason,
      'rejectionReasonHindi': rejectionReasonHindi,
      'uploadedDocuments': uploadedDocuments,
    };
  }

  factory SubsidyApplication.fromJson(Map<String, dynamic> json) {
    return SubsidyApplication(
      id: json['id'],
      schemeId: json['schemeId'],
      farmerName: json['farmerName'],
      farmerId: json['farmerId'],
      mobileNumber: json['mobileNumber'],
      applicationData: json['applicationData'],
      status: json['status'],
      submissionDate: json['submissionDate'] != null
          ? DateTime.parse(json['submissionDate'])
          : DateTime.now(),
      approvalDate: json['approvalDate'] != null
          ? DateTime.parse(json['approvalDate'])
          : null,
      disbursalDate: json['disbursalDate'] != null
          ? DateTime.parse(json['disbursalDate'])
          : null,
      approvedAmount: json['approvedAmount'],
      rejectionReason: json['rejectionReason'],
      rejectionReasonHindi: json['rejectionReasonHindi'],
      uploadedDocuments: List<String>.from(json['uploadedDocuments'] ?? []),
    );
  }
}

class EligibilityResult {
  final String schemeId;
  final bool isEligible;
  final double eligibilityScore; // 0-100
  final List<String> eligibleCriteria;
  final List<String> ineligibleCriteria;
  final List<String> eligibleCriteriaHindi;
  final List<String> ineligibleCriteriaHindi;
  final double estimatedBenefit;
  final List<String> recommendations;
  final List<String> recommendationsHindi;

  EligibilityResult({
    required this.schemeId,
    required this.isEligible,
    required this.eligibilityScore,
    required this.eligibleCriteria,
    required this.ineligibleCriteria,
    required this.eligibleCriteriaHindi,
    required this.ineligibleCriteriaHindi,
    required this.estimatedBenefit,
    required this.recommendations,
    required this.recommendationsHindi,
  });

  Map<String, dynamic> toJson() {
    return {
      'schemeId': schemeId,
      'isEligible': isEligible,
      'eligibilityScore': eligibilityScore,
      'eligibleCriteria': eligibleCriteria,
      'ineligibleCriteria': ineligibleCriteria,
      'eligibleCriteriaHindi': eligibleCriteriaHindi,
      'ineligibleCriteriaHindi': ineligibleCriteriaHindi,
      'estimatedBenefit': estimatedBenefit,
      'recommendations': recommendations,
      'recommendationsHindi': recommendationsHindi,
    };
  }

  factory EligibilityResult.fromJson(Map<String, dynamic> json) {
    return EligibilityResult(
      schemeId: json['schemeId'],
      isEligible: json['isEligible'],
      eligibilityScore: json['eligibilityScore'],
      eligibleCriteria: List<String>.from(json['eligibleCriteria']),
      ineligibleCriteria: List<String>.from(json['ineligibleCriteria']),
      eligibleCriteriaHindi: List<String>.from(json['eligibleCriteriaHindi']),
      ineligibleCriteriaHindi:
          List<String>.from(json['ineligibleCriteriaHindi']),
      estimatedBenefit: json['estimatedBenefit'],
      recommendations: List<String>.from(json['recommendations']),
      recommendationsHindi: List<String>.from(json['recommendationsHindi']),
    );
  }
}

class SchemeNotification {
  final String id;
  final String title;
  final String titleHindi;
  final String message;
  final String messageHindi;
  final String type; // new_scheme, deadline_reminder, status_update, general
  final String? schemeId;
  final DateTime createdAt;
  final DateTime? expiryDate;
  final bool isRead;
  final bool isImportant;

  SchemeNotification({
    required this.id,
    required this.title,
    required this.titleHindi,
    required this.message,
    required this.messageHindi,
    required this.type,
    this.schemeId,
    required this.createdAt,
    this.expiryDate,
    this.isRead = false,
    this.isImportant = false,
  });

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'titleHindi': titleHindi,
      'message': message,
      'messageHindi': messageHindi,
      'type': type,
      'schemeId': schemeId,
      'createdAt': createdAt.toIso8601String(),
      'expiryDate': expiryDate?.toIso8601String(),
      'isRead': isRead,
      'isImportant': isImportant,
    };
  }

  factory SchemeNotification.fromJson(Map<String, dynamic> json) {
    return SchemeNotification(
      id: json['id'],
      title: json['title'],
      titleHindi: json['titleHindi'],
      message: json['message'],
      messageHindi: json['messageHindi'],
      type: json['type'],
      schemeId: json['schemeId'],
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      expiryDate: json['expiryDate'] != null
          ? DateTime.parse(json['expiryDate'])
          : null,
      isRead: json['isRead'] ?? false,
      isImportant: json['isImportant'] ?? false,
    );
  }
}
